src/Amazonka/S3/StreamingUpload.hs

```diff
@@ -283,34 +284,43 @@ data S = S !Builder {-# UNPACK #-} !Int
 newS :: S
 newS = S mempty 0
 
-newSFrom :: ByteString -> S
-newSFrom bs = S (byteStringCopy bs) (B.length bs)
-
 appendS :: S -> ByteString -> S
 appendS (S builder len) bs = S (builder <> byteStringCopy bs) (len + B.length bs)
 
 finaliseS :: S -> IO ByteString
 finaliseS (S builder builderLen) = do
   fptr <- mallocForeignPtrBytes builderLen
   let ptr = unsafeForeignPtrToPtr fptr
-      bufWriter = runBuilder builder
-  bufWriter ptr builderLen >>= \case
+  runBuilder builder ptr builderLen >>= \case
     (written, Done)
       | written == builderLen -> pure $! PS fptr 0 builderLen
       | otherwise ->
           error $ "finaliseS: bytes written didn't match, expected: " <> show builderLen <> " got: " <> show written
     (_written, _) -> error "Something went very wrong"
 
--- Right means the buffer needs more data to fill it
--- Left means the buffer is full
-processChunk :: ChunkSize -> ByteString -> S -> IO (Either S S)
+-- @Right@ means the buffer needs more data to fill it
+-- @Left@ means the buffer is full so it should be yielded, and returns the
+-- remainder of the last input processed.
+processChunk :: ChunkSize -> ByteString -> S -> Either (S,ByteString) S
```

Comment:
Since you're looking to minimise allocations in your other PR, what if you used `B.take` instead of `B.splitAt` and returned an offset back into your original ByteString?

Bonus points for dependent type shenanigans to ensure that the offset you pass around is within the input `ByteString`.

Reply:
That's basically just a more complicated way of doing the same thing - splitAt basically is just returning two offsets into the same bytestring, and IIRC the offset is unboxed into the bytestring constructor. It's noise in terms of allocations compared to the buffers.

Reply:
Oh, that's pretty cool.

Reply:
https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/src/Data.ByteString.html#splitAt